File: Medium/46_Permutations/Solution.py
from typing import List, Optional


class Solution:
    def getPermute(self, nums: list, idx: int, ans: list):
        if (idx == len(nums)):
            ans.append(nums[:])
            return
        
        for i in range(idx, len(nums)):
            nums[idx], nums[i] = nums[i], nums[idx]
            self.getPermute(nums, idx+1, ans)
            nums[idx], nums[i] = nums[i], nums[idx]

    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.getPermute(nums, 0, ans)
        return ans


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    print(solution.permute([1, 2, 3]))

    # test cases 2
    print(solution.permute([0,1]))

    # test cases 3
    print(solution.permute([1]))

File: Medium/46_Permutations/Solution.java
import java.util.ArrayList;
import java.util.List;


public class Solution {
    private static void getPermute(int[] nums, List<Integer> ds, boolean[] isValid, List<List<Integer>> ans) {
        if (ds.size() == nums.length) {
            List<Integer> list = new ArrayList<>();

            for (int i = 0; i < ds.size(); i++) {
                list.add(ds.get(i));
            }
            ans.add(list);
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (isValid[i] == false) {
                ds.add(nums[i]);
                isValid[i] = true;
                getPermute(nums, ds, isValid, ans);
                isValid[i] = false;
                ds.remove(ds.size() - 1);
            }
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        List<Integer> ds = new ArrayList<>();
        boolean[] isValid = new boolean[nums.length];
        List<List<Integer>> ans = new ArrayList<>();
        
        getPermute(nums, ds, isValid, ans);
        return ans;
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        int[] nums1 = {1, 2, 3};
        List<List<Integer>> ans1 = solution.permute(nums1);
        for (List<Integer> l : ans1) {
            System.out.println(l);
        }

        // test cases 2
        int[] nums2 = {1, 2, 3};
        List<List<Integer>> ans2 = solution.permute(nums2);
        for (List<Integer> l : ans2) {
            System.out.println(l);
        }
    }
}
